metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_profiler.py

```diff
@@ -0,0 +1,117 @@
+        for field_id, valueEncoded in manifestValues.items():
+            field = schema.find_field(field_id)
+            valueDecoded = Conversions.from_byte_buffer(field.type, valueEncoded)
+            if valueDecoded:
+                fieldPath = fieldPaths.get(field_id)
+                if not fieldPath:
+                    raise RuntimeError(f"Failed to find fieldPath for field_id {field_id}")
+                agg = aggregatedValues.get(fieldPath)
+                aggregatedValues[fieldPath] = aggregator(agg, valueDecoded) if agg else valueDecoded
+
+    def profileTable(self, env: str, dataset_name: str, table: Table, schemaMetadata: SchemaMetadata) -> MetadataWorkUnit:
+        rowCount = int(table.current_snapshot().summary["total-records"])
+        dataset_profile = DatasetProfileClass(
+            timestampMillis=get_sys_time(),
+            rowCount=rowCount,
+            columnCount=len(schemaMetadata.fields),
+            fieldProfiles=[]
+            # columnCount=len(table.schema().columns())
+        )
+
+        fieldPaths: dict[int, str] = table.schema()._id_to_name #dict[int, str] = {v: k for k, v in index_by_name(table.schema()).items()}
+        current_snapshot: Snapshot = table.current_snapshot()
+        totalCount = 0
+        nullCounts = {}
+        valueCounts = {}
+        minBounds = {}
+        maxBounds = {}
+        manifest: ManifestFile
+        try:
+            for manifest in current_snapshot.manifests:
+                manifest_input_file = FileSystemInputFile.from_location(manifest.manifest_path, table.ops.conf)
+                manifest_reader = ManifestReader.read(manifest_input_file)
+                dataFile: DataFile
+                for dataFile in manifest_reader.iterator():
+                    self._aggregateCounts(nullCounts, fieldPaths, dataFile.null_value_counts())
+                    self._aggregateCounts(valueCounts, fieldPaths, dataFile.value_counts())
+                    self._aggregateBounds(table.schema(), min, minBounds, fieldPaths, dataFile.lower_bounds())
+                    self._aggregateBounds(table.schema(), max, maxBounds, fieldPaths, dataFile.upper_bounds())
+                    totalCount += dataFile.record_count()
+        # TODO Work on error handling to provide better feedback.  Iceberg exceptions are weak...
+        except FileSystemNotFound as e:
+            raise Exception(f"Error loading table manifests") from e
+        if rowCount is not None and rowCount > 0:
+            # Iterating through fieldPaths introduces unwanted stats for list element fields...
+            for fieldPath in fieldPaths.values():
+                column_profile = DatasetFieldProfileClass(fieldPath=fieldPath)
+                column_profile.nullCount = nullCounts.get(fieldPath, 0)
+                column_profile.nullProportion = column_profile.nullCount / rowCount
+                # column_profile.uniqueCount = valueCounts.get(fieldPath, 0)
```

Comment:
Why has the uniqueCount been commented out?

Reply:
Because it doesn't work unfortunately.  I kept it there as a reminder to explain why and here we are.  In Iceberg, a table is composed of multiple data files and for each there, there is an associated manifest.  The role of that manifest is to keep metadata that will be used by a plan to determine if the engine should open that data file or not.  I'm using those metrics to profile the data.  One of those metrics is valueCount, which by definition is (from Iceberg spec):
`Map from column id to number of values in the column (including null and NaN values)`
So its the number of values (distinct values) for **that data file**, not the table...  

Here is an example to illustrate this:
We have a table with 2 data file:
DataFile A, column `X` has values: `A, B, C, A, B`
So its value_counts map will have `X: 3`, so 3 distinct values for column `X`.

DataFile B, column `X` has values: `A, D, E, E, F`
So its value_counts map will have `X: 4`, so 4 distinct values for column `X`.

If I do the sum, I end up with 7 distinct values, but in fact I only have 6 distinct values in the table...

Reply:
Wow that's a pain... maybe we could leave a comment in case someone works out a solution in the future?

Reply:
I can certainly do that, yes.  I bounced this off to one of the Iceberg dev that I know and he confirmed that there isn't much we can do for the distinct count.  We would need to read the data file, which is more expensive.  We could use the sum and get an approximation of the total number of distinct values, but it wouldn't be accurate and we could end up with large % (a table of 200 rows could have 600 distinct values if the data is spread over many data files...).

One way to complete those stats would be to leverage a tool like Apache Spark...  it's just unfortunate to have to rely on another technology...  😢